virtcontainers/pkg/annotations/annotations.go

```diff
@@ -8,6 +8,15 @@ package annotations
 const (
 	vcAnnotationsPrefix = "com.github.containers.virtcontainers."
 
+	// SandboxMemory is a sandbox annotation for passing a per sandbox memory limit
+	// This detemines the memory size of the virtual machine
+	// Typically an admission controller would set this. This allows hypervisors that do not support
+	// memory hotplug to have a VM that is sized to match the sum of the container resources
+	SandboxMemory = vcAnnotationsPrefix + "SandboxMemory"
+
+	// SandboxCpus is a sandbox annotation for passing a per sandbox CPU limit
+	SandboxCpus = vcAnnotationsPrefix + "SandboxCpus"
```

Comment:
I think these should be:

```
com.github.containers.virtcontainers.sandbox.memory
com.github.containers.virtcontainers.sandbox.cpus
```

or, from #753:

```
io.kata-containers.config.virtcontainers.sandbox.memory
io.kata-containers.config.virtcontainers.sandbox.cpus
```

Reply:
@jodh-intel based on @bergwolf  original proposal we should be directly able to map to the configuration.toml variables right. So should it be

```
io.kata-containers.config.hypervisor.default_vcpus
io.kata-containers.config.hypervisor.default_memory
io.kata-containers.config.hypervisor.enable_debug

```
or
```
io.kata-containers.config.hypervisor.firecracker.default_vcpus
io.kata-containers.config.hypervisor.firecracker.default_memory
io.kata-containers.config.hypervisor.firecracker.enable_debug
```

The hypervisor vs hypervisor.firecracker is a bit confusing. Which one should we use.
I like `io.kata-containers.config.hypervisor.default_vcpus` but then we lose the direct mapping.

That way we have a full and direct mapping for each and every config option?

Same with

`proxy.kata` vs `proxy` and `shim.kata` vs `shim`

Reply:
My only concern was not wanting `Foo${blah}` but instead `foo.${blah}`, so this sounds good to me.

Reply:
@bergwolf @jodh-intel Which one

```
io.kata-containers.config.hypervisor.default_vcpus
```

or 
```
io.kata-containers.config.hypervisor.firecracker.default_vcpus
io.kata-containers.config.hypervisor.qemu.default_vcpus
```

I prefer the first method as it is but more generic and also due to the fact that virtcontainers/hypervisor.go etc are generic, but the second method is an exact mapping.

Reply:
My preference is the first:
> ```io.kata-containers.config.hypervisor.default_vcpus```

Reply:
@egernst I tend to agree with you but we need to define `CommonHypervisorConfig` first otherwise it will be pain to manually convert each field to a hypervisor type specific one.

Reply:
At the risk of over-engineering, we could support a generic default:

```
io.kata-containers.config.hypervisor.default.default_vcpus
```
(that "double default" is ugly so maybe `io.kata-containers.config.hypervisor.default.vcpus` or `io.kata-containers.config.hypervisor.generic.default_vcpus`).

... and a hypervisor-specific value which, if set, takes priority over `default`:

```
io.kata-containers.config.hypervisor.firecracker.vcpus
io.kata-containers.config.hypervisor.qemu.vcpus
```

As long as there are no hypervisors called `default`, that should work ;)

Reply:
And of course whichever strategy we decide upon needs to be documented: I think in an actual `README.md` (not just the code).

Reply:
@jodh-intel I like default. Let me create a first cut implementation with the same.

Reply:
Any update @mcastelino?

Reply:
I am revisiting this feature to push an implementation for this. Looking at how things are implemented in Kata today, here are my thoughts:
The main areas that the end-user would need custom configurations through annotations are Hypervisor, Agent and Runtime(maybe in that order).
For hypervisors, we have a generic structure in the code today called `HypervisorConfig` for the hypervisor configuration, instead of a per hypervisor config structures specific to say qemu, firecracker etc.
So, I think we can start with supporting hypervisor annotations in the form 
`io.kata-containers.config.hypervisor.default_vcpus` 
that would apply to the chosen hypervisor for that workload. This would fit it pretty well with the existing code as well.

If we do want to support overrides in the form of per hypervisor specific annotations  and we see that is something that the users do want, we can consider adding hypervisor specific annotations of the form that @jodh-intel suggested:
`io.kata-containers.config.hypervisor.firecracker.vcpus`
I propose we tackle this in the second stage of this feature, if we really do see value in this.

While, we do want to support all the configurations that we currently have in our toml file, I am going to propose to tackle a subset of the configurations for the first pass.
I will document what these are, when I push the code for the initial implementation.

In case of the agent, it would make sense to follow the same generic approach, although if we look at the code for the agent config, we do not have a generic `AgentConfig`, but a specific `KataAgentConfig`, which is ok since technically we support just the kata-agent today.

I would also like to add that historically, we have supported a limited subset of annotations for custom kernel, rootfs etc of the form: 
```
com.github.containers.virtcontainers.KernelPath: "/boot/vmlinuz-custom-container"
```
See the existing annotation prefix `com.github.containers.virtcontainers`:
https://github.com/kata-containers/runtime/blob/2ed94cbd9d9090ae85fa2dc8b39cfaa46ddf8f29/virtcontainers/pkg/annotations/annotations.go#L9

I plan to drop that prefix and replace that with the one discussed here and make sure the existing annotations comply with the hierarchical format discussed here.
I think it should be ok to drop backward compatibility in this case, as this would mean a simple change in the pod yaml.
Let me know what others think. 
cc @bergwolf @jodh-intel @egernst